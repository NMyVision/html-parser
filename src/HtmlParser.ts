import type HtmlNodeAttributes from "./HtmlNodeAttributes.ts";
import HtmlNode from "./HtmlNode.ts";
import { defaultParserOptions, type ParserOptions } from "./ParserOptions.ts";
import StringQueue from "./StringQueue.ts";

class KeyValuePair<K, T> {
  constructor(public key: K, public value?: T) {
  }
}
  /**
   * @module
   *
   * This module contains functions to search the database.
   *
   * @example
   * ```ts
   * import { HtmlParser } from "@nmyvision/HtmlParser";
   *
   * const hp = new HtmlParser();
   * const result = hp.parse(`<article id="editor"/>`);
   * ```
   */
export class HtmlParser {
  q: StringQueue;
  source: string;

  get index(): number {
    return this.q.index;
  }

  /** */
  constructor(public options: ParserOptions = defaultParserOptions) {
    this.source = "";
    this.q = new StringQueue();
  }

  /**
   * This function parses the html string and returns a list of nodes.
   *
   * @param source the html string to be parsed
   * @returns HTMLNode[]
   */
  public parse(source: string): HtmlNode[] {
    this.source = source;
    this.q = new StringQueue(source);

    return this.internalParse();
  }

  private internalParse(parent?: HtmlNode): HtmlNode[] {
    const { q } = this;
    const list: HtmlNode[] = [];
    let node: HtmlNode = HtmlNode.empty;

    while (q.any()) {
      if (q.peek(2) == "</") {
        q.dequeue(2);
        if (parent != null) {
          parent.endTag = q.run(">");
        } else if (node != null) {
          node.endTag = q.run(">");
        } else {
          q.run(">");
        }
        q.dequeue();
        return list;
      } else if (q.peek(4) == "<!--") {
        const comment = this.getComment();
        node = HtmlNode.CreateComment();
        node.content = comment;
        list.push(node);
      } else if (q.peek(2) == "<!") {
        this.q.dequeue(2);
        const content = this.q.run(">");
        this.q.dequeue();

        node = HtmlNode.CreateDocType(content);
        list.push(node);
      } else if (q.peek() == "<") {
        node = HtmlNode.CreateElement(this.index);

        q.dequeue();
        node.tag = this.getTagName();

        if (q.peek() === " " || q.peek() === "\r" || q.peek() === "\n") {
          node.attributes = this.getAttributes();
        }

        q.skipSpace();

        if (q.isEmpty()) {
          node.setEndIndex(this.index, this.source);
          list.push(node);
          break;
        }

        // instantly closed
        if (q.peek() == ">" && (q.peek(3) == "></")) {
          q.dequeue(3);
          node.endTag = q.run(">");
          q.dequeue();

          node.setEndIndex(this.index, this.source);
          list.push(node);
        } // self closing element
        else if (q.peek(2) == "/>") {
          node.endTag = q.run("/>");
          q.dequeue(2);
          node.selfClosing = true;

          node.setEndIndex(this.index, this.source);
          list.push(node);
        } // self closing tags that don't have '/>' ie: <br>
        else if (
          q.peek() == ">" && this.options.selfClosingTags.includes(node.tag)
        ) {
          q.dequeue();
          node.selfClosing = true;
          node.setEndIndex(this.index, this.source);
          list.push(node);
        } else if (/script/i.test(node.tag)) {
          q.dequeue();
          node.content = q.run("</script");
          node.endTag = "script";
          list.push(node);
        }
      } else if (q.peek() == ">") {
        q.dequeue();
        node.addChildren(this.internalParse(node));
        node.setEndIndex(this.index, this.source);
        list.push(node);
      } else {
        let text = "";

        while (q.any()) {
          text += q.dequeue();
          if (q.any() && q.peek() == "<") break;
        }
        if (text.trim().length === 0) {
          if (this.options.ignoreWhitespace === false) {
            list.push(HtmlNode.CreateWhiteSpace(text));
          }
        } else {
          list.push(HtmlNode.CreateText(text));
        }
      }
    }

    return list;
  }

  getAttributes(): HtmlNodeAttributes {
    const GetAttribute = (): KeyValuePair<string, string> => {
      const name = this.q.run("=", " ", ">");

      if (q.peek() == " ") return new KeyValuePair<string, string>(name);

      this.q.dequeue();

      if (q.any()) {
        // attr=value is valid so check for the scenerio
        if (q.peek() == "'" || q.peek() == '"') {
          const del = this.q.dequeue();
          const value = this.q.run(del);
          this.q.dequeue();
          return new KeyValuePair<string, string>(name, value);
        } else {
          const value = this.q.run(" ", ">", "<", "'", '"', "=", "`");
          return new KeyValuePair<string, string>(name, value);
        }
      }

      return new KeyValuePair<string, string>(name);
    };

    const attrs: HtmlNodeAttributes = {};
    const { q } = this;
    if (q.isEmpty()) return attrs;

    if (q.peek() == "/") return attrs;

    q.skipSpace();

    do {
      const kv = GetAttribute();
      attrs[kv.key] = kv.value; //.set(kv.key, kv.value);
      q.skipSpace();
    } while (q.any() && q.peek() != ">" && q.peek(2) != "/>");

    return attrs;
  }

  getTagName(): string {
    return this.q.run(" ", ">", "/", "\r", "\n");
  }

  getComment(): string {
    this.q.dequeue(4);
    const text = this.q.run("-->");
    this.q.dequeue(3);
    return text;
  }
}
