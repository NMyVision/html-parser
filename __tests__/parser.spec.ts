import { describe, test } from "jsr:@std/testing/bdd";
import { expect } from "jsr:@std/expect";
import { HtmlParser } from "../src/HtmlParser.ts";
import { assertSnapshot } from "@std/testing/snapshot";

describe("HtmlParser Tests", () => {
  test("simple test", async (t) => {
    const hp = new HtmlParser();

    const html = "<div />";

    const result = hp.parse(html)[0];

    await assertSnapshot(t, result);

    expect(result.tag).toEqual("div");

    expect(result.outerHTML).toEqual(html);
  });

  test("attribute test", async (t) => {
    const hp = new HtmlParser();

    const html = "<input disabled required type='text' />";

    const result = hp.parse(html)[0];

    await assertSnapshot(t, result);

    expect(result.tag).toEqual("input");

    expect(Object.keys(result.attributes as object).length).toEqual(3);

    expect(result.attributes?.type).toEqual("text");

    expect(result.outerHTML).toEqual(html);
  });

  test("attribute test", async (t) => {
    const hp = new HtmlParser();

    const innerHtml = `<input disabled required type='text' />`;

    const html = `<div>${innerHtml}</div>`;

    let result = hp.parse(html)[0];

    await assertSnapshot(t, result);

    expect(result.tag).toEqual("div");

    expect(Object.keys(result.attributes as object).length).toEqual(0);

    expect(result.outerHTML).toEqual(html);

    result = result.children[0];

    expect(result.tag).toEqual("input");

    expect(Object.keys(result.attributes as object).length).toEqual(3);

    expect(result.attributes?.type).toBeTruthy();

    expect(result.attributes?.type).toEqual("text");

    expect(result.outerHTML).toEqual(innerHtml);
  });

  test("incomplete sample", async (t) => {
    const hp = new HtmlParser();

    let result = hp.parse(`<article id="edito`);

    await assertSnapshot(t, result);

    expect(result[0].tag).toEqual("article");

    expect(result[0]?.attributes?.id).toEqual("edito");

    result = hp.parse(`<article id="editor"`);

    expect(result[0].tag).toEqual("article");

    expect(result[0]?.attributes?.id).toEqual("editor");
  });

  test("vue sample", () => {
    const vue = `<article id="editor">
  <textarea :value="input" @input="update"></textarea>
  <div v-html="compiledMarkdown"></div>
</article>`;

    const hp = new HtmlParser();

    const result = hp.parse(vue);

    let el = result[0];
    expect(el.tag).toEqual("article");

    el = result[0].children[0];

    expect(el.tag).toEqual("textarea");
    expect(Object.values(el.attributes as object).length).toEqual(2);

    el = result[0].children[1];
    expect(el.tag).toEqual("div");
    expect(Object.values(el.attributes as object).length).toEqual(1);
  });

  test("vue multiline sample", () => {
    const vue = `<article id="editor">
  <textarea
    :value="input"
    @input="update => alert('Ok')"
    >Some Content</textarea>
  <div v-html="compiledMarkdown"></div>
</article>`;

    const hp = new HtmlParser();

    const result = hp.parse(vue);

    let el = result[0];
    expect(el.tag).toEqual("article");

    el = result[0].children[0];

    expect(el.tag).toEqual("textarea");
    expect(el.children[0].content).toEqual("Some Content");
    expect(Object.values(el.attributes as object).length).toEqual(2);

    el = result[0].children[1];
    expect(el.tag).toEqual("div");
    expect(Object.values(el.attributes as object).length).toEqual(1);
  });

  test("jsx sample", async (t) => {
    const jsx = `<fieldset>
        <legend>Enter temperature in {scaleNames[scale]}:</legend>
        <input value={temperature}
               onChange={this.handleChange} />
      </fieldset>`;

    const hp = new HtmlParser();

    const result = hp.parse(jsx);

    await assertSnapshot(t, result);
  });

  test("index.html sample", () => {
    const html = `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <link rel="icon" href="/favicon.ico" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Vite App</title>
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@tailwindcss/ui@latest/dist/tailwind-ui.min.css">
<style>
html {
  --scrollbarBG: #ebeff1;
  --thumbBG: #CFD8DC;
}
</style>
</head>
<!-- this is a comment -->
<body class="min-h-screen">
  <div id="app" class="min-h-screen"></div>
  <script type="module" src="/src/main.js"></script>
</body>
</html>`;

    const hp = new HtmlParser();

    const result = hp.parse(html);

    expect(result[0].type).toEqual(0);

    expect(result[1].type).toEqual(3);

    JSON.stringify(result);
  });
});
