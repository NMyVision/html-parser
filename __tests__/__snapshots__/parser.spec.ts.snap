export const snapshot = {};

snapshot[`HtmlParser Tests > simple test 1`] = `
HtmlNode {
  _outerHTML: "<div />",
  attributes: {
    "/": undefined,
  },
  children: [],
  content: undefined,
  endIndex: 7,
  endTag: undefined,
  selfClosing: undefined,
  startIndex: 0,
  tag: "div",
  type: 3,
}
`;

snapshot[`HtmlParser Tests > attribute test 1`] = `
HtmlNode {
  _outerHTML: "<input disabled required type='text' />",
  attributes: {
    disabled: undefined,
    required: undefined,
    type: "text",
  },
  children: [],
  content: undefined,
  endIndex: 39,
  endTag: "",
  selfClosing: true,
  startIndex: 0,
  tag: "input",
  type: 3,
}
`;

snapshot[`HtmlParser Tests > attribute test 2`] = `
HtmlNode {
  _outerHTML: "<div><input disabled required type='text' /></div>",
  attributes: {},
  children: [
    HtmlNode {
      _outerHTML: "<input disabled required type='text' />",
      attributes: {
        disabled: undefined,
        required: undefined,
        type: "text",
      },
      children: [],
      content: undefined,
      endIndex: 44,
      endTag: "",
      selfClosing: true,
      startIndex: 5,
      tag: "input",
      type: 3,
    },
  ],
  content: undefined,
  endIndex: 50,
  endTag: "div",
  selfClosing: undefined,
  startIndex: 0,
  tag: "div",
  type: 3,
}
`;

snapshot[`HtmlParser Tests > incomplete sample 1`] = `
[
  HtmlNode {
    _outerHTML: '<article id="edito',
    attributes: {
      dito: "ito",
      id: "edito",
      ito: "to",
      o: undefined,
      to: "o",
    },
    children: [],
    content: undefined,
    endIndex: 18,
    endTag: undefined,
    selfClosing: undefined,
    startIndex: 0,
    tag: "article",
    type: 3,
  },
]
`;

snapshot[`HtmlParser Tests > jsx sample 1`] = `
[
  HtmlNode {
    _outerHTML: "<fieldset>
        <legend>Enter temperature in {scaleNames[scale]}:</legend>
        <input value={temperature}
               onChange={this.handleChange} />
      </fieldset>",
    attributes: {},
    children: [
      HtmlNode {
        _outerHTML: "<legend>Enter temperature in {scaleNames[scale]}:</legend>",
        attributes: {},
        children: [
          HtmlNode {
            _outerHTML: undefined,
            attributes: {},
            children: [],
            content: "Enter temperature in {scaleNames[scale]}:",
            endIndex: 0,
            endTag: undefined,
            selfClosing: undefined,
            startIndex: 0,
            tag: undefined,
            type: 4,
          },
        ],
        content: undefined,
        endIndex: 77,
        endTag: "legend",
        selfClosing: undefined,
        startIndex: 19,
        tag: "legend",
        type: 3,
      },
      HtmlNode {
        _outerHTML: "<input value={temperature}
               onChange={this.handleChange} />",
        attributes: {
          onChange: "{this.handleChange}",
          value: "{temperature}
",
        },
        children: [],
        content: undefined,
        endIndex: 159,
        endTag: "",
        selfClosing: true,
        startIndex: 86,
        tag: "input",
        type: 3,
      },
    ],
    content: undefined,
    endIndex: 177,
    endTag: "fieldset",
    selfClosing: undefined,
    startIndex: 0,
    tag: "fieldset",
    type: 3,
  },
]
`;
